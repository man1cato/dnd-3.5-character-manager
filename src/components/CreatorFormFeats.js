import React, { useState, useEffect } from 'react'
import _ from 'lodash'
import { ErrorMessage } from 'formik'
import FeatModal from './FeatModal'
import Selector from './Selector'

const CreatorFormFeats = ({
	values, 
	feats, 
	setFieldValue, 
	validateForm
}) => {
	const [selectedFeatIds, setSelectedFeatIds] = useState(values.feats)
	const [selected, setSelected] = useState(undefined)

	useEffect(() => {
		setFieldValue('feats', selectedFeatIds)
	}, [selectedFeatIds])

	useEffect(() => {
		validateForm()
	}, [])

	const Content = ({selected}) => (
		<div>
			<div className="form-group--35"><b>Type(s):</b> <div>{selected.types.join(", ")}</div></div>
			<div className="form-group--35 align-top"><b>Description:</b> <div>{selected.description}</div></div>
		</div>
	)
	
	return (
		<div>
			<h3 className="row--center">Select Feats</h3>

			<Selector
				apiObject={feats} 
				fieldName="feats"
				Content={Content}
				selectedObjIds={selectedFeatIds}
				setSelectedObjIds={setSelectedFeatIds}
			/>

			<div className="divider"></div>

			<h4 className="row--left">Selected Feats:</h4>
			{_.map(values.feats, (featId) => {
				const feat = {id: featId, ...feats[featId]}
				return (
					<div className="form-group--80" key={featId}>
						<button
							className="button--link"
							type="button"
							id={featId}
							onClick={() => setSelected(feat)}
						>
							{feat.name}
						</button>
									
						<button className="button" type="button" onClick={() => setSelectedFeatIds(_.without(selectedFeatIds, featId))}>x</button>
					</div>
				)
			})}

			<FeatModal
				selected={selected}
				handleCloseModal={() => setSelected(undefined)}
			/>

			<ErrorMessage className="row--left form-group--error" name="feats" component="div" />
		</div>
	)
}
export default CreatorFormFeats