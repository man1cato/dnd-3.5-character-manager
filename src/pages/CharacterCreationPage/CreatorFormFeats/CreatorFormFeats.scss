@import '../CharacterCreationPage.scss';


.CreatorFormFeats__list {
   display: flex;
   justify-content: space-between;
   align-items: center;
   margin-bottom: $m-size;
   
   > :first-child {
      flex: 1 1 80%;
      padding: 0 $s-size 0 0;
   }
   > :last-child {
      flex: 1 0 100px;   
   }
}