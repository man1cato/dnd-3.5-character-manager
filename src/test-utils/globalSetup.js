import fs from 'fs'
import getApi from '../utils/getApi'

export default async () => {
   const api = await getApi()
   
   fs.writeFile('src/test-utils/fixtures/api.json', JSON.stringify(api), (err) => {
      if (err) throw err
   })
}