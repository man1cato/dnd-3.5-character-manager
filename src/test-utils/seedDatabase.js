import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}