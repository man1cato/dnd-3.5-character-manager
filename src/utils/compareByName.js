export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}