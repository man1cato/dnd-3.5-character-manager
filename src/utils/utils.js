import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])