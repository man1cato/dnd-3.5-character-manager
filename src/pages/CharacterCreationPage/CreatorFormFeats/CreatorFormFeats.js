import React, { useState, useEffect } from 'react'
import _ from 'lodash'
import { ErrorMessage } from 'formik'

import FeatModal from '../../../components/Modals/FeatModal'
import Selector from '../../../components/Selector/Selector'
import { apiObjectToArray } from '../../../utils/utils'
import './CreatorFormFeats.scss'


const CreatorFormFeats = ({
	values, 
	feats, 
	setFieldValue 
}) => {
	const featsArray = apiObjectToArray(feats)
	const [availableFeats, setAvailableFeats] = useState(featsArray)
	const [selectedFeatIds, setSelectedFeatIds] = useState(values.feats)
	const [clickedFeat, setClickedFeat] = useState(undefined)

	const handleSelect = selectedFeatId => {
		setSelectedFeatIds([...selectedFeatIds, selectedFeatId])
	}

	useEffect(() => {
		setAvailableFeats(_.filter(featsArray, feat => 
			!_.includes(selectedFeatIds, feat.id)
		))
		setFieldValue('feats', selectedFeatIds)
	}, [selectedFeatIds])

	const SelectorContent = ({ currentItemId }) => {
		const currentFeat = feats[currentItemId]
		return (
			<div className="section">
				<p><b>Type(s):</b> {currentFeat.types.join(", ")}</p>
				<p><b>Description:</b> {currentFeat.description}</p>
			</div>
		)
	}
	
	return (
		<div className="container--body">						
			<h3 className="section__title">Select Feats</h3>

			<Selector
				items={availableFeats} 
				Content={SelectorContent}
				handleSelect={handleSelect}
			/>

			<div className="divider"></div>

			<h4>Selected Feats:</h4>
			{_.map(values.feats, featId => {
				const feat = { id: featId, ...feats[featId] }
				return (
					<div className="CreatorFormFeats__list" key={featId}>
						<button
							className="button--link"
							type="button"
							id={featId}
							onClick={() => setClickedFeat(feat)}
						>
							{feat.name}
						</button>
									
						<button 
							className="button" 
							type="button" 
							onClick={() => setSelectedFeatIds(_.without(selectedFeatIds, featId))}
						>
							<ion-icon name="close" size="small" />
						</button>
					</div>
				)
			})}

			<FeatModal
				clickedFeat={clickedFeat}
				handleCloseModal={() => setClickedFeat(undefined)}
			/>

			<ErrorMessage className="CreatorForm__error" name="feats" component="div" />
		</div>
	)
}


export default CreatorFormFeats