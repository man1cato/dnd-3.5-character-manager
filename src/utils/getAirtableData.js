import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}