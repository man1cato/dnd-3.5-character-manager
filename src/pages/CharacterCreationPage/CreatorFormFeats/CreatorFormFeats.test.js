import React from 'react'
import _ from 'lodash'

import { apiData, renderWithRedux, FormikWrapper } from '../../../test-utils/utils'
import CreatorFormFeats from './CreatorFormFeats'


const setFieldValue = jest.fn()
let props

beforeAll(async () => {
   const api = await apiData()
   props = {
      values: { feats: [] },
      feats: _.omitBy(api.feats, (feat) => !!feat.prerequisites || _.includes(feat.types, 'Epic')),
      setFieldValue
   }
})


test('should correctly render CreatorFormFeats', () => {
   const { container } = renderWithRedux(<CreatorFormFeats {...props} />, null, FormikWrapper)
   expect(container.firstChild).toMatchSnapshot()
})