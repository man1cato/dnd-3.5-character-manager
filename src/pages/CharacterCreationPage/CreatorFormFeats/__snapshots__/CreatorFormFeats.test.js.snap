// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render CreatorFormFeats 1`] = `
<form>
  <div
    class="container--body"
  >
    <h3
      class="section__title"
    >
      Select Feats
    </h3>
    <div
      class="Selector__row"
    >
      <select
        class="Selector__dropdown select"
      >
        <option
          value="recdmotqOlGgN2otb"
        >
          Ability Focus
        </option>
        <option
          value="recoTlzPQ37bNLFsm"
        >
          Acrobatic
        </option>
        <option
          value="rec8uIUF1rFAa2VuG"
        >
          Agile
        </option>
        <option
          value="recvzMqdCBIzd7z9n"
        >
          Alertness
        </option>
        <option
          value="recnrQnRgUIJQkJOz"
        >
          Animal Affinity
        </option>
        <option
          value="recRXzmlbn1MzfoKr"
        >
          Armor Proficiency (Light)
        </option>
        <option
          value="recOCyJT5oWscbchC"
        >
          Athletic
        </option>
        <option
          value="recQyW6ByR4ClPt80"
        >
          Autonomous
        </option>
        <option
          value="recjk6hkPLOgMt6eb"
        >
          Blind-Fight
        </option>
        <option
          value="rec6j8vDL2r3twF8o"
        >
          Cloak Dance
        </option>
        <option
          value="reclpbMRlP2bTvA8K"
        >
          Colossal Wild Shape
        </option>
        <option
          value="recdR990pCoRoCtHN"
        >
          Combat Casting
        </option>
        <option
          value="recrc1dWL5N3RPKd6"
        >
          Combat Expertise
        </option>
        <option
          value="recXnkMszr2a4cMG2"
        >
          Combat Reflexes
        </option>
        <option
          value="recI7IQe71f44n0qs"
        >
          Deadly Precision
        </option>
        <option
          value="recFrheLpwNhWFzS3"
        >
          Deceitful
        </option>
        <option
          value="recPlnmL0xLwCkhqW"
        >
          Deft Hands
        </option>
        <option
          value="rec83JXxeqFaq13F1"
        >
          Diligent
        </option>
        <option
          value="rechBgTFWqu5YnOc1"
        >
          Diminutive Wild Shape
        </option>
        <option
          value="reccKhXP1P03au3WT"
        >
          Divine Might
        </option>
        <option
          value="recrA225bMNGuLGF2"
        >
          Divine Vengeance
        </option>
        <option
          value="recyqJvG85XCUZG4x"
        >
          Dodge
        </option>
        <option
          value="rec35lBICs5ipkw7m"
        >
          Dragon Wild Shape
        </option>
        <option
          value="recHnrZHpmp7psvj2"
        >
          Empower Spell-Like Ability
        </option>
        <option
          value="recVCgn24xVSXigbM"
        >
          Endurance
        </option>
        <option
          value="rec1UNqoxirOSxnno"
        >
          Eschew Materials
        </option>
        <option
          value="recfeBOXTX2up9aUs"
        >
          Exotic Weapon Proficiency
        </option>
        <option
          value="recGByPLOPCGQZEW1"
        >
          Extra Music
        </option>
        <option
          value="recMrrf3flxHUSgMR"
        >
          Extra Turning
        </option>
        <option
          value="recuXAYNPNco1uMdG"
        >
          Eyes in the Back of Your Head
        </option>
        <option
          value="recjTsJRbGi8Lptpa"
        >
          Fine Wild Shape
        </option>
        <option
          value="recw2WZMbQ3gonDew"
        >
          Flyby Attack
        </option>
        <option
          value="rectdd0HKAYm9SmIx"
        >
          Gargantuan Wild Shape
        </option>
        <option
          value="recboOAfWHQTIZGfr"
        >
          Great Fortitude
        </option>
        <option
          value="recEepS7K6IXvQOQU"
        >
          Hover
        </option>
        <option
          value="recEpNvwwJ71UvJm5"
        >
          Improved Counterspell
        </option>
        <option
          value="rec5slnNRWWJC8fb2"
        >
          Improved Critical
        </option>
        <option
          value="recPwR8YqatQuJqXS"
        >
          Improved Elemental Wild Shape
        </option>
        <option
          value="recQDaucsIOGOt9hp"
        >
          Improved Familiar
        </option>
        <option
          value="reciazjHVWQmiq1jP"
        >
          Improved Initiative
        </option>
        <option
          value="recJSbPjUKOTWNe6m"
        >
          Improved Natural Armor
        </option>
        <option
          value="recNxmdwGsIHTAPxj"
        >
          Improved Natural Attack
        </option>
        <option
          value="recZ7I3tsw8Wglevc"
        >
          Improved Spell Resistance
        </option>
        <option
          value="recuEsiEZ1O2huYCr"
        >
          Improved Turning
        </option>
        <option
          value="recCmnimojxd0MNLS"
        >
          Improved Unarmed Strike
        </option>
        <option
          value="recs5pgQMU6zotYqL"
        >
          Investigator
        </option>
        <option
          value="rec0Rmpc0o4zr5KQZ"
        >
          Iron Will
        </option>
        <option
          value="recBS1Vlwp32LWNNe"
        >
          Jack of All Trades
        </option>
        <option
          value="recaVxDe7IV4Ibppd"
        >
          Leadership
        </option>
        <option
          value="recHPtIzQdLipkdLi"
        >
          Lightning Reflexes
        </option>
        <option
          value="rec43FVgR2bo1YDR6"
        >
          Magical Aptitude
        </option>
        <option
          value="recExtvt4xN6i7AiO"
        >
          Magical Beast Companion
        </option>
        <option
          value="recq2EZvZ93PkGmh4"
        >
          Magical Beast Wild Shape
        </option>
        <option
          value="recu6KSoLrL5T47wp"
        >
          Martial Weapon Proficiency
        </option>
        <option
          value="recUv7HQzFXwLgRB9"
        >
          Mind Over Body
        </option>
        <option
          value="reciJuIqTWUQRhYv0"
        >
          Mounted Combat
        </option>
        <option
          value="recKkMD7JxJW1XkHj"
        >
          Multiattack
        </option>
        <option
          value="recvJkbFXbH5ZXAXa"
        >
          Multiweapon Fighting
        </option>
        <option
          value="reccG488HS3PmSuGK"
        >
          Natural Spell
        </option>
        <option
          value="rec5982GoHsn1tfis"
        >
          Negative Energy Burst
        </option>
        <option
          value="recoQZX87YxSgmV2T"
        >
          Negotiator
        </option>
        <option
          value="recL3E07iXqlgd6Mv"
        >
          Nimble Fingers
        </option>
        <option
          value="recFp03E0YFKaq5xr"
        >
          Open Minded
        </option>
        <option
          value="recPjLbGCj3xRNPU3"
        >
          Persuasive
        </option>
        <option
          value="reco0bwyops62ZxPn"
        >
          Plant Defiance
        </option>
        <option
          value="rec81hWf4GNmLBenB"
        >
          Plant Wild Shape
        </option>
        <option
          value="rec0Aa93zpU5V3Asy"
        >
          Point Blank Shot
        </option>
        <option
          value="recuv5RdScK7Vqhf5"
        >
          Power Attack
        </option>
        <option
          value="recpHwtjV89aVIoGB"
        >
          Psionic Affinity
        </option>
        <option
          value="recLp3uvkYGZGyywW"
        >
          Quick Draw
        </option>
        <option
          value="rec6QsVR3a7aEc8Ev"
        >
          Quicken Spell-Like Ability
        </option>
        <option
          value="recmNzyXLWdoJ94Vp"
        >
          Rapid Metabolism
        </option>
        <option
          value="recEk5kUnIREk59lF"
        >
          Rapid Reload
        </option>
        <option
          value="rec3Z5EqrUQXIcAjy"
        >
          Reckless Offense
        </option>
        <option
          value="rec02pDGILJ5EZz1W"
        >
          Run
        </option>
        <option
          value="recBV4PEpBYb8mVC0"
        >
          Self-Sufficient
        </option>
        <option
          value="recm19w48xltockBW"
        >
          Shield Proficiency
        </option>
        <option
          value="rec2wriPyA1jtpZih"
        >
          Simple Weapon Proficiency
        </option>
        <option
          value="recaj8lyaZepMlFCM"
        >
          Skill Focus
        </option>
        <option
          value="recpeswIYdGxGtd4g"
        >
          Snatch
        </option>
        <option
          value="recr648zasZ6BHY5D"
        >
          Spell Focus
        </option>
        <option
          value="recQM5J2Z4Bg345O0"
        >
          Spell Penetration
        </option>
        <option
          value="recB7JU0g2Bfz5xqS"
        >
          Stand Still
        </option>
        <option
          value="rece4EQ3n0Ya33LGf"
        >
          Stealthy
        </option>
        <option
          value="reczbG3R5q7p4lj9R"
        >
          Toughness
        </option>
        <option
          value="recxkSuS8IvkzRVyV"
        >
          Track
        </option>
        <option
          value="recePeBndah0BTNs4"
        >
          Two-Weapon Fighting
        </option>
        <option
          value="recJFvWvKKUkM6NJM"
        >
          Undead Mastery
        </option>
        <option
          value="reccZ5xjJ9eWUxvY8"
        >
          Vermin Wild Shape
        </option>
        <option
          value="recVdnKmMFz54fNLF"
        >
          Weapon Finesse
        </option>
        <option
          value="recy5YNMs3lo9rYTp"
        >
          Weapon Focus
        </option>
        <option
          value="recF2gHjeFMKLdUmE"
        >
          Wild Talent
        </option>
        <option
          value="recI40vCPdTE5LsEu"
        >
          Wingover
        </option>
      </select>
      <button
        class="Selector__button button"
        type="button"
      >
        Select
      </button>
    </div>
    <div
      class="section"
    >
      <p>
        <b>
          Type(s):
        </b>
         
        General, Creature
      </p>
      <p>
        <b>
          Description:
        </b>
         
      </p>
    </div>
    <div
      class="divider"
    />
    <h4>
      Selected Feats:
    </h4>
  </div>
</form>
`;
